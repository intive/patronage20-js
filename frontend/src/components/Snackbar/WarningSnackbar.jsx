
import React, { useEffect, useState } from 'react'

import { Snackbar, SnackbarContent, Slide, IconButton } from '@material-ui/core'
import { Warning, Close } from '@material-ui/icons'
import { amber } from '@material-ui/core/colors'
import { makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles(theme => ({
  warning: {
    backgroundColor: amber[700]
  },
  icon: {
    fontSize: 20,
    opacity: 0.9,
    marginRight: theme.spacing(1)
  },
  message: {
    display: 'flex',
    alignItems: 'center'
  }
}))

// eslint-disable-next-line react/prop-types
const WarningSnackbar = ({ pingEndpoint }) => {
  const [open, setOpen] = useState(false)
  const classes = useStyles()
  const showSnackbar = (error) => {
    error.response.status === 408
      ? setOpen(true)
      : console.error(error)
  }

  const checkConnection = function () {
    if (!navigator.onLine) {
      setOpen(true)
    } else {
      pingEndpoint()
        .then(() => setOpen(false))
        .catch(showSnackbar)
    }
  }

  let interval
  useEffect(() => {
    clearInterval(interval)
    checkConnection()
    interval = setInterval(checkConnection, 10000)

    return () => {
      setOpen(false)
      clearInterval(interval)
    }
  }, [])

  useEffect(() => window.addEventListener('offline', () => setOpen(true)))
  useEffect(() => window.addEventListener('online', () => setOpen(false)))

  const handleClose = () => setOpen(false)

  return (
    <Snackbar
      data-testid='snackbar'
      open={open}
      TransitionComponent={(props) => <Slide {...props} direction='down' />}
      anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
    >
      <SnackbarContent
        className={classes.warning}
        message={
          <span className={classes.message} id='client-snackbar'>
            <Warning className={classes.icon} />
            Hej, coś nie styka! Sprawdź połączenie!
          </span>
        }
        action={[
          <IconButton
            key='close'
            aria-label='Close'
            color='inherit'
            onClick={handleClose}
          >
            <Close />
          </IconButton>
        ]}
      />
    </Snackbar>
  )
}

export default WarningSnackbar
