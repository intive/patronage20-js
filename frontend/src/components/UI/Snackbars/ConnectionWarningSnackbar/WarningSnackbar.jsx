import { useEffect } from 'react'
import { useSnackbar } from 'notistack'
import { useTranslation } from 'react-i18next'

const WarningSnackbar = ({ pingEndpoint }) => {
  const { enqueueSnackbar, closeSnackbar } = useSnackbar()
  const { t } = useTranslation()

  let key
  const status = [408, 502, 504]

  const handleSnackbarEnqueueing = () => {
    if (!key) {
      key = enqueueSnackbar(t('connection-warning'), {
        variant: 'warning',
        persist: true,
        onExiting: () => { key = null }
      })
    }
  }

  const handleSnackbarClosing = () => {
    if (key) {
      closeSnackbar(key)
      key = null
    }
  }

  const showSnackbar = (error) => {
    !error.response || status.includes(error.response.status)
      ? handleSnackbarEnqueueing()
      : console.error(error)
  }

  const checkConnection = function () {
    if (!navigator.onLine) {
      handleSnackbarEnqueueing()
    } else {
      pingEndpoint()
        .then(() => handleSnackbarClosing())
        .catch(showSnackbar)
    }
  }

  let interval
  useEffect(() => {
    checkConnection()
    interval = setInterval(() => checkConnection(), 10000)
    return () => {
      handleSnackbarClosing()
      clearInterval(interval)
    }
  }, [])

  useEffect(() => window.addEventListener('offline', () => handleSnackbarEnqueueing()))
  useEffect(() => window.addEventListener('online', () => handleSnackbarClosing()))

  return null
}

export default WarningSnackbar
