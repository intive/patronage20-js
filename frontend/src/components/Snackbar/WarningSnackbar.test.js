import React from 'react'
import { cleanup, render, waitForElement } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import '@testing-library/jest-dom/matchers'
import { create } from 'react-test-renderer'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import WarningSnackbar from './index'

const axiosMock = new MockAdapter(axios)

const pingEndpointMock = () => axios.get('/health-check', {
  timeout: 5000
})

let isOnlineGetter

describe.only('WarningSnackbar component', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    isOnlineGetter = jest.spyOn(window.navigator, 'onLine', 'get')
  })

  afterEach(async () => {
    jest.clearAllTimers()
    jest.useRealTimers()
    await cleanup()
  })

  it('Matches the snapshot', () => {
    const snackbar = create(<WarningSnackbar />)
    expect(snackbar.toJSON()).toMatchSnapshot()
  })

  it('should ping the health-check endpoint after 10 second and still be hidden', (done) => {
    const callMock = jest
      .fn()
      .mockReturnValue([200, {}])
      .mockName('serverEndpoint')
    axiosMock.onGet('/health-check')
      .reply(callMock)
    const { container } = render(<WarningSnackbar pingEndpoint={pingEndpointMock} />)
    jest.advanceTimersByTime(20100)
    setImmediate(() => {
      expect(callMock).toHaveBeenCalledTimes(3)
      expect(container.firstChild).toBeNull()
      done()
    })
  })

  it('should open the snackabar when the HTTP code is 408', async () => {
    const callMock = jest.fn()
      .mockReturnValueOnce([200, {}])
      .mockReturnValueOnce([408, {}])
      .mockName('serverEndpoint')
    axiosMock
      .onGet('/health-check')
      .reply(callMock)
    const { getByTestId } = render(<WarningSnackbar pingEndpoint={pingEndpointMock} />)
    jest.advanceTimersByTime(11000)
    await waitForElement(() => getByTestId('snackbar'))
    expect(callMock).toHaveBeenCalledTimes(2)
  })

  it('should open the snackabar when there is a network error', async () => {
    isOnlineGetter.mockReturnValue(false)
    axiosMock.onGet('/health-check')
    const { getByTestId } = render(<WarningSnackbar pingEndpoint={pingEndpointMock} />)
    await waitForElement(() => getByTestId('snackbar'))
  })
})
