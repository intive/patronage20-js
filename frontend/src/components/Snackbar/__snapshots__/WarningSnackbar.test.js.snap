// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarningSnackbar component Matches the snapshot 1`] = `null`;
